cml-tdengine/src/core/inference.rs

```diff
@@ -31,105 +31,108 @@ impl<D: IntoDsn + Clone> Inference<Field, Value, i64, Manager<TaosBuilder>> for
-                    Some(value) => ColumnView::from(value.clone()),
-                    None => null.clone(),
-                };
-                let mut values = vec![ColumnView::from_nanos_timestamp(vec![current_ts]), output];
-                if let Some(fields) = sample.optional_fields() {
-                    values.append(
-                        &mut fields
-                            .iter()
-                            .map(|f| ColumnView::from(f.to_owned()))
-                            .collect::<Vec<ColumnView>>(),
-                    )
+                let mut values_to_bind = Vec::<Vec<ColumnView>>::new();
+                for sample in data {
+                    let output = match sample.output() {
+                        Some(value) => ColumnView::from(value.clone()),
+                        None => null.clone(),
+                    };
+                    let mut values =
+                        vec![ColumnView::from_nanos_timestamp(vec![current_ts]), output];
+                    if let Some(fields) = sample.optional_fields() {
+                        values.append(
+                            &mut fields
+                                .iter()
+                                .map(|f| ColumnView::from(f.to_owned()))
+                                .collect::<Vec<ColumnView>>(),
+                        )
+                    }
+
+                    values_to_bind.push(values);
+                    current_ts += Duration::from_nanos(1).as_nanos() as i64;
                 }
 
-                values_to_bind.push(values);
-                current_ts += Duration::from_nanos(1).as_nanos() as i64;
-            }
+                batch_state.remove(&metadata.batch);
+                cvar.notify_one();
+                values_to_bind
+            };
+            let field_placeholder = get_placeholders(values_to_bind.first().unwrap());
+            stmt.prepare(&format!(
+                "INSERT INTO ? USING inference TAGS ({}) VALUES ({})",
+                tag_placeholder, field_placeholder
+            ))
+            .await?;
 
-            batch_state.remove(&metadata.batch);
-            cvar.notify_one();
-            values_to_bind
+            stmt.set_tbname_tags(metadata.batch(), &tags).await?;
+            for values in values_to_bind {
+                stmt.bind(&values).await?;
+            }
+            stmt.add_batch().await?;
+            stmt.execute().await?;
         };
-        let field_placeholder = get_placeholders(values_to_bind.first().unwrap());
-        stmt.prepare(&format!(
-            "INSERT INTO ? USING inference TAGS ({}) VALUES ({})",
-            tag_placeholder, field_placeholder
-        ))
-        .await?;
-
-        stmt.set_tbname_tags(metadata.batch(), &tags).await?;
-        for values in values_to_bind {
-            stmt.bind(&values).await?;
-        }
-        stmt.add_batch().await?;
-        stmt.execute().await?;
         Ok(())
```

Comment:
To satisfy the signature requirement of `Future<Output = Result<()>>`, the returned value should be included in the async block. The async function is a special case of an async block and can produce a `Future` object.

It is important to note that since all methods on `TDengine` are declared with its reference (`&self`), the references of `D` may need to be shared between threads. Therefore, it may be necessary to mark it with `Sync`.